history_and_modals.md

```diff
@@ -1,158 +1,357 @@
+
+### Realistic examples
+
+The above sections give illustrative usage of the API. The following ones show how the API could be incorporated into realistic apps and UI components.
+
+#### A sidebar
+
+For a sidebar (e.g. behind a hamburger menu), which wants to hide itself on a user-provided close signal, that could be hooked up as follows:
+
+```js
+hamburgerMenuButton.addEventListener('click', () => {
+  const watcher = new ModalCloseWatcher();
+
+  sidebar.animate([{ transform: 'translateX(-200px)' }, { transform: 'translateX(0)' }]);
+
+  watcher.onclose = () => {
+    sidebar.animate([{ transform: 'translateX(0)' }, { transform: 'translateX(-200px)' }]);
+  };
+
+  // TODO: also call watcher.destroy() if the sidebar closes for other reasons,
+  // e.g. clicking outside the sidebar.
+});
 ```
 
-The `onclose` handler is constructed similar to `unbeforeunload` event to ensure
-that a "save data" prompt cannot be abused as an escape hatch to block the back
-button.
+#### A picker
+
+For a "picker" control that wants to close itself on a user-provided close signal, code like the following would work:
 
-If Dialog API cannot be extended in this way, there could be a similar new
-HTML element that can support all of these features and allow flexible styling.
-The `dialog` element itself could extend from this new type of the element.
+```js
+class MyPicker extends HTMLElement {
+  #button;
+  #overlay;
+  #watcher;
 
-### A back button event propagated via the Esc key
+  constructor() {
+    super();
+    this.#button = /* ... */;
 
-The browser could automatically propagate the back button activation as an <kbd>Esc</kbd>
-key event. The main benefit of this would be that many existing popup/dialog/dropdown
-implementations that already support the <kbd>Esc</kbd> key, could automatically
-gain the back button support without any changes or only with minor updates.
+    this.#overlay = /* ... */;
+    this.#overlay.hidden = true;
+    this.#overlay.querySelector('.close-button').addEventListener('click', () => {
+      this.#watcher.destroy();
+    });
 
-The use of such API could then be as simple as keyboard event handling:
+    this.#button.onclick = () => {
+      this.overlay.hidden = false;
 
-```javascript
-const popup = document.createElement('div');
-document.onkeydown = (e) => {
-  if (e.code === 'Escape') {
-    if (!allDataSaved()) {
-      return 'Are you sure you want to close?';
+      this.#watcher = new ModalCloseWatcher();
+      this.#watcher.onclose = () => this.overlay.hidden = true;
     }
-    popup.remove();
   }
-};
+}
 ```
 
-Similarly, this event handler looks like `unbeforeunload` to ensure that this
-feature cannot be abused as an escape hatch to block the back button. Consequently,
-the `event.preventDefault()` can cancel the "keydown" event, but it has no
-effect on the back button behavior.
+_TODO: include a React example._
```

Comment:
I don't know enough about React, thus this TODO :). Would you be able to say more about what's fragile? Are you saying that the current proposal is very fragile, or instead that a declarative alternative would be fragile?

Reply:
There are a couple of nuances here:

1. The `open` prop would be propagated and rerendered by React asynchronously. I'd expect the time between a user clicking the "open" button of some sort and this code executing would be very small - most likely the next event loop via `setTimeout(..., 0)`. However, it depends on how user gesture "corridor" is defined by a particular browser. The last I remember there were a lot of differences between Chrome/FF/Safari. But I do think this would be fine in Chrome.
2. Very react-specific, it's usually hard to tell in React what happens on the first render vs Nth render in reaction to some event. I guess that's the whole point of React. So this code would try to execute `new ModalWatcher()` even when the popup is rendered open by default w/o a user action.

All of this is very similar to `<dialog open>` (declarative) vs `dialog.showModal()` (imperative). So, nothing "weird" about it. But it'd be good if the API was ok with all this. For instance, we could stipulate that the `new ModalWatcher()` never fails to create, but will ignore "close" events if it wasn't created based on the user gesture.

Reply:
I think that's pretty much the same as

>  In particular, it would be great if a single ModalCloseWatcher could be created without user activation.